service/matching/task_queue_partition_manager.go

```diff
@@ -96,51 +103,103 @@ func newTaskQueuePartitionManager(
+		autoEnableRateLimiter: quotas.NewRateLimiter(1.0/60, 1),
+	}
+	pm.initCtx, pm.initCancel = context.WithCancel(context.Background())
 
 	if pm.partition.IsRoot() {
 		pm.cache = cache.New(10000, &cache.Options{
 			TTL: max(1, tqConfig.TaskQueueInfoByBuildIdTTL())}, // ensure TTL is never zero (which would disable TTL)
 		)
 	}
 
+	return pm, nil
+}
+
+var errDefaultQueueNotInit = serviceerror.NewInternal("defaultQueue is not initializaed")
+
+func (pm *taskQueuePartitionManagerImpl) initialize() {
 	unload := func(bool) {
 		pm.unloadFromEngine(unloadCauseConfigChange)
 	}
 
-	var fairness bool
-	fairness, pm.cancelFairnessSub = tqConfig.EnableFairnessSub(unload)
-	// Fairness is disabled for sticky queues for now so that we can still use TTLs.
-	tqConfig.EnableFairness = fairness && partition.Kind() != enumspb.TASK_QUEUE_KIND_STICKY
-	if fairness {
-		tqConfig.NewMatcher = true
-	} else {
-		tqConfig.NewMatcher, pm.cancelNewMatcherSub = tqConfig.NewMatcherSub(unload)
+	err := pm.userDataManager.WaitUntilInitialized(pm.initCtx)
+	if err != nil {
+		pm.defaultQueueFuture.Set(nil, err)
+		return
+	}
+	data, _, err := pm.getPerTypeUserData()
+	if err != nil {
+		pm.defaultQueueFuture.Set(nil, err)
+		return
+	}
+
+	pm.autoEnable = pm.config.AutoEnable()
+	pm.fairnessState = data.GetFairnessState()
+	switch {
+	case !pm.autoEnable || pm.fairnessState == persistencespb.FAIRNESS_STATE_UNSPECIFIED:
+		var fairness bool
+		fairness, pm.cancelFairnessSub = pm.config.EnableFairnessSub(unload)
+		// Fairness is disabled for sticky queues for now so that we can still use TTLs.
+		pm.config.EnableFairness = fairness && pm.partition.Kind() != enumspb.TASK_QUEUE_KIND_STICKY
+		if fairness {
+			pm.config.NewMatcher = true
+		} else {
+			pm.config.NewMatcher, pm.cancelNewMatcherSub = pm.config.NewMatcherSub(unload)
+		}
+	case pm.fairnessState == persistencespb.FAIRNESS_STATE_V0:
+		pm.config.NewMatcher = false
+		pm.config.EnableFairness = false
+	case pm.fairnessState == persistencespb.FAIRNESS_STATE_V1:
+		pm.config.NewMatcher = true
+		pm.config.EnableFairness = false
+	case pm.fairnessState == persistencespb.FAIRNESS_STATE_V2:
+		pm.config.NewMatcher = true
+		if pm.partition.Kind() == enumspb.TASK_QUEUE_KIND_STICKY {
+			pm.config.EnableFairness = false
+		} else {
+			pm.config.EnableFairness = true
+		}
+	default:
+		pm.defaultQueueFuture.Set(nil, serviceerror.NewInternal("Unknown FairnessState in UserData"))
+		return
 	}
 
-	defaultQ, err := newPhysicalTaskQueueManager(pm, UnversionedQueueKey(partition))
+	defaultQ, err := newPhysicalTaskQueueManager(pm, UnversionedQueueKey(pm.partition))
 	if err != nil {
-		return nil, err
+		pm.defaultQueueFuture.Set(nil, err)
+		return
 	}
-	pm.defaultQueue = defaultQ
-	return pm, nil
+	defaultQ.Start()
+	pm.defaultQueueFuture.Set(defaultQ, nil)
+}
+
+func (pm *taskQueuePartitionManagerImpl) defaultQueue() physicalTaskQueueManager {
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
```

Comment:
I kinda hate that we have to this (I did it somewhere else too). this is our own future implementation, I think we should just add a GetIfReady to it.